engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java

```diff
@@ -0,0 +1,836 @@
+          return handleIncomparableValues(x, y);
+        } else if (isTrue(xLessThanYRes)) {
+          return ascending ? -1 : 1;
+        } else {
+          // Check if x > y
+          Object yLessThanXRes = lessThanNode.execute(y, x);
+          if (isTrue(yLessThanXRes)) {
+            return ascending ? 1 : -1;
+          } else {
+            // yLessThanXRes is either Nothing or False
+            return handleIncomparableValues(y, x);
+          }
+        }
+      }
+    }
+
+    private int handleIncomparableValues(Object x, Object y) {
+      switch (problemBehavior) {
+        case REPORT_ERROR -> throw new CompareException(x, y);
+        case REPORT_WARNING -> attachIncomparableValuesWarning(x, y);
+      }
+      if (isPrimitiveValue(x) || isPrimitiveValue(y)) {
+        if (isPrimitiveValue(x) && isPrimitiveValue(y)) {
+          return handleIncomparablePrimitives(x, y);
+        } else if (isPrimitiveValue(x)) {
+          // Primitive values are always before non-primitive values - Default_Comparator
+          // group should be the first one.
+          return ascending ? -1 : 1;
+        } else if (isPrimitiveValue(y)) {
+          return ascending ? 1 : -1;
+        } else {
+          throw new IllegalStateException("Should not be reachable");
+        }
+      } else {
+        // Values other than primitives are compared just by their type's FQN.
+        var xTypeName = getQualifiedTypeName(x);
+        var yTypeName = getQualifiedTypeName(y);
+        return xTypeName.compareTo(yTypeName);
+      }
+    }
+
+    /**
+     * Incomparable primitive values have either different builtin types, or are Nothing or NaN. All
+     * these cases are handled specifically - we hardcode the order of these incomparable values.
+     */
+    private int handleIncomparablePrimitives(Object x, Object y) {
+      int xCost = getPrimitiveValueCost(x);
+      int yCost = getPrimitiveValueCost(y);
+      int res = Integer.compare(xCost, yCost);
+      return ascending ? res : -res;
+    }
+
+    private boolean isPrimitiveValue(Object object) {
+      return isBuiltinType(typeOfNode.execute(object));
+    }
+
+    private String getQualifiedTypeName(Object object) {
+      var typeObj = typeOfNode.execute(object);
+      return toTextNode.execute(typeObj).toString();
+    }
+
+    private int getPrimitiveValueCost(Object object) {
+      if (interop.isNull(object)) {
+        return 200;
+      } else if (isNan(object)) {
+        return 100;
+      } else {
+        var type = typeOfNode.execute(object);
+        return getBuiltinTypeCost(type);
+      }
+    }
+  }
+
+  /**
+   * Comparator for any values. This comparator compares the values by calling back to Enso (by
+   * {@link #compareFunc}), rather than using compare nodes (i.e. {@link LessThanNode}). directly,
+   * as opposed to {@link DefaultSortComparator}.
+   */
+  private final class GenericSortComparator extends SortComparator {
```

Comment:
nitpick: This file is pretty big. Wouldn't it make it more readable to move this class to a separate file?

Reply:
I am afraid not. The inner classes are not `static` on purpose - they use the reference to the outer class.

Reply:
Oh! I missed this fact. Then ignore this.